### DISCUSSION TITLE
Custom initialization for nn layers?

### DISCUSSION THREAD

**Discussion by theunnecessarythings** [huggingface/candle#1831] [Q&A] [answered]
Do we have a way to do custom weight initialization for nn layers (conv, linear, norms etc). For eg. in tch-rs the configs had ws_init and bs_init. From what I have seen from the code, these are initialized locally in the functions/modules (for candle). Am I missing something?

**Comment by LaurentMazare** [comment, marked as answer]
We usually provide two ways to create nn layers, e.g. via `conv1d` that would perform the initialization if required or via `Conv1d::new` that would take the weight tensor as inputs. `conv1d` calls `Conv1d::new` and you can easily duplicate the related code to perform a different initialization if you want, e.g. just copy [this function](https://github.com/huggingface/candle/blob/0c5eecbc0faa7e642210800c735ad8137d5a9e08/candle-nn/src/conv.rs#L323) and tweak the "kaiming normal" bit to your liking.

**Reply by theunnecessarythings** [reply]
I was gonna do the same as well. Just wanted to make sure I wasn't missing anything since tch had some option to do this via config itself. Thanks for the fast reply.

**Reply by laptou** [reply]
Is there a way to do an initialization besides the ones in the `Init` struct? I want to initialize a parameter using `linspace`/`arange`, unless it already exists in the `VarBuilder`, but I can't figure out how to do that.

**Reply by LaurentMazare** [reply]
An easy way though not very efficient is to initialize it using one of the built-in initializers, and then set it to the value that you want (either in the same module or at the level of your training script).
A more complicated way would be to write your own var-builder to handle the new initialization but not sure it's worth the hassle.

**Reply by laptou** [reply]
But if I set it to a certain value, then won't that overwrite the value if it was present in the `VarBuilder`? Essentially, I want my vector to have a default value if I'm initializing my model to the default state, but this tensor's value can change over time. If I'm loading a pretrained version of this model then I want the trained value, not the original `linspace` default

**Reply by LaurentMazare** [reply]
Right, it would indeed overwrite the value assuming that it's a variable and not a tensor, but you can probably control from your code whether you want to do the `set` or not. It might be as simple as checking whether the tensor is a variable or not, a variable comes from a `VarMap` (used for training) and a tensor from the other varbuilders. You just have to be a bit cautious if you start loading some checkpoint (i.e. use a `VarMap` and get the initial values from a safetensor file).

**Reply by laptou** [reply]
Ah, I didn't know about `Tensor::is_variable`. Thanks!
